ibverbs/src/lib.rs

```diff
@@ -2021,6 +2160,89 @@ impl QueuePair {
         self._post_one_sided(local, remote, wr_id, opcode, imm_data)
     }
 
+    #[inline]
+    /// Remote RDMA write with EFA.
+    /// immediate data can be used to signal the completion of the write operation
+    /// the other side uses post_recv on a dummy buffer and get the imm data from the work completion
+    pub fn post_write_efa(
+        &mut self,
+        local: &[LocalMemorySlice],
+        remote_mr: RemoteMemorySlice,
+        remote_endpoint: QueuePairEndpoint,
+        wr_id: u64,
+        imm_data: Option<u32>,
+    ) -> io::Result<()> {
+        let qp_ex = unsafe { ffi::ibv_qp_to_qp_ex(self.qp) };
```

Comment:
check that the result is not nullptr, then dereference once and use below, then you likely dont need the unsafe after this line

Reply:
hmm, not able to remove the full unsafe blocks., Since inner functions like `wr_rdma_write_imm` is unsafe as well.